### PR TITLE
RASS-303 create court cases add a sentence fine amount

### PR DESCRIPTION
**Pull request by pavilionsahota**

### CODE REVIEW

File: server/@types/models/index.d.ts [hunk 1]
```diff
@@ -43,6 +43,7 @@ declare module 'models' {
     sentence?: Sentence
     terrorRelated?: boolean
     legacyData: Record<string, never>
+    fineAmount?: number
```
Comment (carloveo-moj) [inline server/@types/models/index.d.ts:46]: A fineAmount is associated to a Sentence not an Offence, can this be moved to the Sentence object below

File: server/routes/offenceRoutes.ts [hunk 2]
```diff
@@ -786,6 +799,73 @@ export default class OffenceRoutes {
     )
   }
 
+  public getFineAmount: RequestHandler = async (req, res): Promise<void> => {
```
Comment (carloveo-moj) [inline server/routes/offenceRoutes.ts:802]: There is a journey to get to this page via the edit offence screen, look in other pages to add a `submitToEditOffence` query to this page

File: server/routes/offenceRoutes.ts [hunk 3]
```diff
@@ -786,6 +799,73 @@ export default class OffenceRoutes {
...
+      nomsId,
+      courtCaseReference,
+      appearanceReference,
+      addOrEditCourtCase,
+      addOrEditCourtAppearance,
+      offenceFineAmountForm,
+      offenceReference,
+      isAddOffences: this.isAddJourney(addOrEditCourtCase, addOrEditCourtAppearance),
+      errors: req.flash('errors') || [],
+    })
```
Comment (carloveo-moj) [inline server/routes/offenceRoutes.ts:830]: there is no back link defined here

File: server/routes/offenceRoutes.ts [hunk 4]
```diff
@@ -786,6 +799,73 @@ export default class OffenceRoutes {
...
+      nomsId,
+      courtCaseReference,
+      offenceFineAmountForm,
+    )
+
+    if (errors.length > 0) {
+      req.flash('errors', errors)
+      req.flash('offenceFineAmountForm', { ...offenceFineAmountForm })
+      return res.redirect(
+        `/person/${nomsId}/${addOrEditCourtCase}/${courtCaseReference}/${addOrEditCourtAppearance}/${appearanceReference}/offences/${offenceReference}/fine-amount`,
```
Comment (carloveo-moj) [inline server/routes/offenceRoutes.ts:855]:
```suggestion
        `/person/${nomsId}/${addOrEditCourtCase}/${courtCaseReference}/${addOrEditCourtAppearance}/${appearanceReference}/offences/${offenceReference}/fine-amount${submitToEditOffence ? '?submitToEditOffence=true' : ''}`,
```

File: server/routes/offenceRoutes.ts [hunk 5]
```diff
@@ -1490,6 +1575,10 @@ export default class OffenceRoutes {
     this.saveOffenceInAppearance(req, nomsId, courtCaseReference, offenceReference, offence)
   }
 
+  private isAddJourney(addOrEditCourtCase: string, addOrEditCourtAppearance: string): boolean {
+    return addOrEditCourtCase.includes('add') && addOrEditCourtAppearance.includes('add')
```
Comment (carloveo-moj) [inline server/routes/offenceRoutes.ts:1579]:
```suggestion
    return addOrEditCourtCase === 'add-court-case' && addOrEditCourtAppearance === 'add-court-appearance'
```

File: server/services/offenceService.ts [hunk 6]
```diff
@@ -318,6 +319,32 @@ export default class OffenceService {
...
+  setOffenceFineAmount(
+    session: CookieSessionInterfaces.CookieSessionObject,
+    nomsId: string,
+    courtCaseReference: string,
+    offenceFineAmountForm: OffenceFineAmountForm,
+  ) {
+    const errors = validate(
+      offenceFineAmountForm,
+      {
+        fineAmount: 'required|minWholeNumber:0',
```
Comment (carloveo-moj) [inline server/services/offenceService.ts:331]: fine amounts can be to two decimal places I think, not just whole numbers

File: server/@types/models/index.d.ts (hunk 1 above)
Comment (carloveo-moj) [reply server/@types/models/index.d.ts:46]: A sentence can't have multiple offences, one offence to one sentence. The fine amount comes after selecting a Imprisonment in default of a fine sentence type not outcome. The sentence type is associated to the sentence not the offence so therefore the fine amount if associated to the sentence not the offence

Comment (pavilionsahota) [reply server/@types/models/index.d.ts]: I've adjusted this

File: server/routes/offenceRoutes.ts (hunk 2 above)
Comment (pavilionsahota) [reply server/routes/offenceRoutes.ts:802]: I've added this in

File: server/routes/offenceRoutes.ts (hunk 4 above)
Comment (pavilionsahota) [reply server/routes/offenceRoutes.ts]: I've changed this as you've suggested

File: server/routes/offenceRoutes.ts (hunk 5 above)
Comment (pavilionsahota) [reply server/routes/offenceRoutes.ts]: I've changed this as you've suggested

File: server/routes/offenceRoutes.ts (hunk 2 above)
Comment (carloveo-moj) [reply server/routes/offenceRoutes.ts:802]: `submitToEditOffence` needs to be added to the options of the page render too https://github.com/ministryofjustice/hmpps-remand-and-sentencing/blob/0d42b051dfd3afb099cc51dc11333b6f73e9cc7c/server/routes/offenceRoutes.ts#L825

File: server/services/offenceService.ts (hunk 6 above)
Comment (carloveo-moj) [reply server/services/offenceService.ts]: it needs to be two decimal places only, a user shouldn't be able to enter 10.3456 for example

**Review by carloveo-moj — APPROVED** [review]
LGTM

**Pull request merged**